File: Backend/src/main/java/com/SalaryApplication/Services/SalaryBreakdownService.java
package com.SalaryApplication.Services;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.SalaryApplication.Models.Deduction;
import com.SalaryApplication.Models.SalaryBreakdown;
import com.SalaryApplication.Models.User;
import com.SalaryApplication.Models.Enums.DeductionTypeEnum;
import com.SalaryApplication.Repositories.DeductionRepository;
import com.SalaryApplication.Repositories.UserRepository;
import com.SalaryApplication.Services.Interfaces.INIService;
import com.SalaryApplication.Services.Interfaces.ISalaryBreakdownService;
import com.SalaryApplication.Services.Interfaces.IStudentFinanceService;
import com.SalaryApplication.Services.Interfaces.ITaxService;

@Service
public class SalaryBreakdownService implements ISalaryBreakdownService {

    @Autowired
    private UserRepository userRepo;
    @Autowired
    private DeductionRepository deductionRepo;

    @Autowired
    private ITaxService taxService;
    @Autowired
    private INIService niService;
    @Autowired
    private IStudentFinanceService studentFinanceService;

    public SalaryBreakdown getSalaryBreakdown(Integer userId) {
        SalaryBreakdown breakdown = new SalaryBreakdown();

        Optional<User> user = userRepo.findById(userId);
        if (user != null) {
            breakdown = getSalaryBreakdown(user.get());
        }

        return breakdown;
    }

    public SalaryBreakdown getSalaryBreakdown(String userName) {
        SalaryBreakdown breakdown = new SalaryBreakdown();

        Optional<User> user = userRepo.findByUsername(userName);
        if (user != null) {
            breakdown = getSalaryBreakdown(user.get());
        }

        return breakdown;
    }

    public SalaryBreakdown getSalaryBreakdown(User user) {
        SalaryBreakdown breakdown = new SalaryBreakdown();

        breakdown.setName(user.getFirstName() + " " + user.getLastName());
        breakdown.setUser(user);

        breakdown.setMonthlySalary(user.getSalary().divide(BigDecimal.valueOf(12), 2, RoundingMode.UP));

        breakdown.setPension(breakdown.getMonthlySalary()
                .multiply((user.getPensionPercentage().divide(BigDecimal.valueOf(100)))).setScale(2, RoundingMode.DOWN));
        if (user.isPensionSalarySacrifice()) {
            breakdown.setSalarySacrifice(breakdown.getPension());
        }

        this.DeductSalarySacrificeDeductions(breakdown);

        breakdown.setTax(taxService.calculateMonthlyTax(breakdown.getSalaryPostSalarySacrifice()));
        breakdown.setTaxFree(taxService.getMonthlyTaxFree());

        breakdown.setnI(niService.calculateMonthlyNI(breakdown.getSalaryPostSalarySacrifice()));

        breakdown.setStudentFinance(studentFinanceService.calculateMonthlyStudentFinance(
                breakdown.getSalaryPostSalarySacrifice(), niService.GetNIBenefitValue(),
                taxService.GetTaxBenefitValue(), user.getStudentFinancePlan()));

        this.CalculateTakehome(breakdown);

        return breakdown;
    }

    private void DeductSalarySacrificeDeductions(SalaryBreakdown breakdown) {
        List<Deduction> deductions = this.deductionRepo
                .findByType(DeductionTypeEnum.SalarySacrifice);
        BigDecimal deductionsTotal = deductions.stream()
                .map(Deduction::getCost)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        breakdown.setSalarySacrifice(breakdown.getSalarySacrifice().add(deductionsTotal));
    }

    private void CalculateTakehome(SalaryBreakdown breakdown) {
        BigDecimal salary = breakdown.getMonthlySalary();

        BigDecimal takehome = salary.subtract(breakdown.getTax());
        takehome = takehome.subtract(breakdown.getnI());
        takehome = takehome.subtract(breakdown.getStudentFinance());

        if (breakdown.getUser().isPensionSalarySacrifice()) {
            takehome = takehome.subtract(breakdown.getPension());
        }

        // Benefits only deducted at end as to ensure they onyl affec relevant
        // calculations
        takehome = takehome.subtract(niService.GetNIBenefitValue());
        takehome = takehome.subtract(taxService.GetTaxBenefitValue());

        BigDecimal payrollDeductions = this.GetPayrollDeductionValue();

        breakdown.setTakehome(takehome.subtract(payrollDeductions));

        // TODO Add section to deduct costs (IE Rent, bills etc)
        breakdown.setTakehomeAfterBills(breakdown.getTakehome().subtract(BigDecimal.ZERO));
    }

    private BigDecimal GetPayrollDeductionValue() {
        List<Deduction> deductions = this.deductionRepo
                .findByType(DeductionTypeEnum.TaxBenefit);
        BigDecimal deductionsTotal = deductions.stream()
                .map(Deduction::getCost)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        return deductionsTotal;
    }
}

File: backend/src/services/salaryBreakdown.service.ts
import { Deduction } from "../entity/Deduction.class";
import { SalaryBreakdown } from "../entity/SalaryBreakdown.class";
import { DeductionType } from "../enums/DeductionType";
import userService from "../services/user.service";
import taxService from "../services/taxCalculation.service";
import niService from "../services/niCalculation.service";
import studentFinanceService from "../services/studentFinanceCalculation.service";
import { User } from "../entity/User.class";

const getSalaryBreakdownById = async (
  userId: number,
): Promise<SalaryBreakdown> => {
  const user = await userService.getUser(userId);
  return getSalaryBreakdown(user);
};

const getSalaryBreakdownByUsername = async (
  userName: string,
): Promise<SalaryBreakdown> => {
  const user = await userService.getUserByUsername(userName);
  return getSalaryBreakdown(user);
};

const getSalaryBreakdown = async (user: User): Promise<SalaryBreakdown> => {
  const breakdown = new SalaryBreakdown(user);

  breakdown.user = user;
  breakdown.monthlySalary = user.salary / 12;
  breakdown.pension = breakdown.monthlySalary * (user.pensionPercentage / 100);

  if (user.pensionSalarySacrifice) {
    breakdown.salarySacrifice = breakdown.pension;
  }

  await deductSalarySacrificeDeductions(breakdown);

  breakdown.tax = await taxService.calculateMonthlyTax(
    breakdown.getSalaryPostSalarySacrifice(),
  );
  breakdown.taxFree = await taxService.getMonthlyTaxFree();
  breakdown.nI = await niService.calculateMonthlyNI(
    breakdown.getSalaryPostSalarySacrifice(),
  );
  if (user.studentFinancePlan !== null) {
    breakdown.studentFinance =
      await studentFinanceService.calculateMonthlyStudentFinance(
        breakdown.getSalaryPostSalarySacrifice(),
        await niService.getNIBenefitValue(),
        await taxService.getTaxBenefitValue(),
        user.studentFinancePlan,
      );
  }

  await calculateTakehome(breakdown);

  return breakdown;
};

const deductSalarySacrificeDeductions = async (
  breakdown: SalaryBreakdown,
): Promise<void> => {
  const deductionsTotal = await getSalarySacrificeDeductionValue(
    breakdown.user,
  );
  breakdown.salarySacrifice = breakdown.salarySacrifice + deductionsTotal;
};

const calculateTakehome = async (breakdown: SalaryBreakdown): Promise<void> => {
  const salary = breakdown.getSalaryPostSalarySacrifice();
  let takehome =
    salary - breakdown.tax - breakdown.nI - breakdown.studentFinance;

  if (!breakdown.user.pensionSalarySacrifice) {
    takehome = takehome - breakdown.pension;
  }

  takehome =
    takehome -
    (await niService.getNIBenefitValue()) -
    (await taxService.getTaxBenefitValue());

  const payrollDeductions = await getPayrollDeductionValue(breakdown.user);
  breakdown.takehome = takehome - payrollDeductions;
  breakdown.bills = await getBillsDeductionValue(breakdown.user);
  breakdown.savingsAndInvestments =
    await getSavingsAndInvestmentsDeductionValue(breakdown.user);
  breakdown.takehomeAfterBillsAndSavings =
    breakdown.takehome - (breakdown.bills + breakdown.savingsAndInvestments);
};

const getSalarySacrificeDeductionValue = async (
  user: User,
): Promise<number> => {
  const deductions = user.deductions.filter(
    (deduction) => deduction.type === DeductionType.SALARY_SACRIFICE,
  );
  const deductionsTotal = deductions.reduce(
    (runningTotal: number, currentDeduction: Deduction) =>
      runningTotal + currentDeduction.cost,
    0,
  );
  return deductionsTotal;
};

const getPayrollDeductionValue = async (user: User): Promise<number> => {
  const deductions = user.deductions.filter(
    (deduction) => deduction.type === DeductionType.TAX_BENEFIT,
  );
  const deductionsTotal = deductions.reduce(
    (runningTotal: number, currentDeduction: Deduction) =>
      runningTotal + currentDeduction.cost,
    0,
  );
  return deductionsTotal;
};

const getBillsDeductionValue = async (user: User): Promise<number> => {
  const deductions = user.deductions.filter(
    (deduction) => deduction.type === DeductionType.BILL,
  );
  const deductionsTotal = deductions.reduce(
    (runningTotal: number, currentDeduction: Deduction) =>
      runningTotal + currentDeduction.cost,
    0,
  );
  return deductionsTotal;
};

const getSavingsAndInvestmentsDeductionValue = async (
  user: User,
): Promise<number> => {
  const deductions = user.deductions.filter(
    (deduction) => deduction.type === DeductionType.SAVING_AND_INVESTMENT,
  );
  const deductionsTotal = deductions.reduce(
    (runningTotal: number, currentDeduction: Deduction) =>
      runningTotal + currentDeduction.cost,
    0,
  );
  return deductionsTotal;
};

export default { getSalaryBreakdownById, getSalaryBreakdownByUsername };
